File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        sLen, pLen = len(s), len(p)
        res = []

        if sLen < pLen:
            return res
        
        sFreq, pFreq = [0] * 26, [0] * 26

        for i in range(pLen):
            sFreq[ord(s[i]) - ord('a')] += 1
            pFreq[ord(p[i]) - ord('a')] += 1

        if sFreq == pFreq:
            res.append(0)

        for i in range(pLen, sLen):
            sFreq[ord(s[i]) - ord('a')] += 1
            sFreq[ord(s[i - pLen]) - ord('a')] -= 1

            if sFreq == pFreq:
                res.append(i - pLen + 1)

        return res

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;
        int pLen = p.length(), sLen = s.length();

        if(sLen < pLen) return result;

        vector<int> sCount(26, 0);
        vector<int> pCount(26, 0);

        for(int i = 0 ; i < pLen ; i++) {
            sCount[s[i] - 'a']++;
            pCount[p[i] - 'a']++;
        }

        if(sCount == pCount) result.push_back(0);

        for(int i = pLen ; i < sLen ; i++) {
            sCount[s[i] - 'a']++;
            sCount[s[i - pLen] - 'a']--;

            if(pCount == sCount) result.push_back(i - pLen + 1);
        }

        return result;
    }
};